File: VB/ThemeSelector/Helpers/BackgroundHelper.vb
' Developer Express Code Central Example:
' How to use the Theme Selector that is used in DemoCenter
' 
' This sample illustrates how to use our theme selector that is used in our demos
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E3392

Imports System
Imports System.Threading
Imports System.Windows
Imports System.Windows.Threading

Namespace ThemeSelector.Helpers
    Public NotInheritable Class BackgroundHelper

        Private Sub New()
        End Sub

        Private Shared dispatcher_Renamed As Dispatcher

        Public Shared Sub DoInBackground(ByVal backgroundAction As Action, ByVal mainThreadAction As Action)
            DoInBackground(backgroundAction, mainThreadAction, 30)
        End Sub
        Public Shared Sub DoInBackground(ByVal backgroundAction As Action, ByVal mainThreadAction As Action, ByVal milliseconds As Integer)
            Dim thread As New Thread(Sub()
                System.Threading.Thread.Sleep(milliseconds)
                If backgroundAction IsNot Nothing Then
                    backgroundAction()
                End If
                If mainThreadAction IsNot Nothing Then
                    Dispatcher.BeginInvoke(mainThreadAction)
                End If
            End Sub)
            thread.Start()
        End Sub
        Public Shared Sub DoInMainThread(ByVal action As Action)
            Dim done As New AutoResetEvent(False)
            Dispatcher.BeginInvoke(CType(Sub()
                action()
                done.Set()
            End Sub, Action)
           )
            done.WaitOne()
        End Sub
        Public Shared Property Dispatcher() As Dispatcher
            Get
                If dispatcher_Renamed Is Nothing Then
                    dispatcher_Renamed = DefaultDispatcher
                End If
                Return dispatcher_Renamed
            End Get
            Set(ByVal value As Dispatcher)
                dispatcher_Renamed = value
            End Set
        End Property
        Private Shared ReadOnly Property DefaultDispatcher() As Dispatcher
            Get
#If SL Then
                Return Deployment.Current.Dispatcher
#Else
                Return Application.Current.Dispatcher
#End If
            End Get
        End Property
    End Class
End Namespace

File: CS/ThemeSelector/Helpers/BackgroundHelper.cs
﻿// Developer Express Code Central Example:
// How to use the Theme Selector that is used in DemoCenter
// 
// This sample illustrates how to use our theme selector that is used in our demos
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E3392

using System;
using System.Threading;
using System.Windows;
using System.Windows.Threading;

namespace ThemeSelector.Helpers {
    public static class BackgroundHelper {
        static Dispatcher dispatcher;

        public static void DoInBackground(Action backgroundAction, Action mainThreadAction) {
            DoInBackground(backgroundAction, mainThreadAction, 30);
        }
        public static void DoInBackground(Action backgroundAction, Action mainThreadAction, int milliseconds) {
            Thread thread = new Thread(delegate() {
                Thread.Sleep(milliseconds);
                if(backgroundAction != null)
                    backgroundAction();
                if(mainThreadAction != null)
                    Dispatcher.BeginInvoke(mainThreadAction);
            });
            thread.Start();
        }
        public static void DoInMainThread(Action action) {
            AutoResetEvent done = new AutoResetEvent(false);
            Dispatcher.BeginInvoke((Action)delegate() {
                action();
                done.Set();
            });
            done.WaitOne();
        }
        public static Dispatcher Dispatcher {
            get {
                if(dispatcher == null)
                    dispatcher = DefaultDispatcher;
                return dispatcher;
            }
            set { dispatcher = value; }
        }
        static Dispatcher DefaultDispatcher {
            get {
#if SL
                return Deployment.Current.Dispatcher;
#else
                return Application.Current.Dispatcher;
#endif
            }
        }
    }
}
